Support coin-selection of Byron inputs in `balanceTransaction`

Anviking opened the pull request:
- [x] Extend `estimateTxFee` and `estimateNumberOfWitnesses` to also be able to deal with Bootstrap witnesses
- [x] Make `produces balanced transactions or fails` property test both `Shelley` and `Byron` wallets

### Comments

Needed for #3746 

### Issue Number

ADP-2268 / ADP-2613

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs:
```diff
@@ -814,6 +902,7 @@ balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
                     -- Could be made smarter by only padding for the script
                     -- integrity hash when we intend to add one. [ADP-2621]
                     scriptIntegrityHashBytes = 32 + 2
+                    -- FIXME: failures from fixing this are interesting
```
```suggestion
```

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs:
```diff
@@ -278,16 +296,71 @@ instance Buildable (PartialTx era) where
+
+-- FIXME
+instance Eq CoinSelection where
+    a == b = False
```
Maybe defunctionalizing is better... Then maybe we can have

```haskell
-- Easily showable in property tests:
data CoinSelection = Shared | ShelleyVk | ByronVk

-- Wallet.hs
class CanBalance k where
    defaultCoinSelection :: Proxy k -> Proxy ktype -> CoinSelection
```
but not sure 🤔

jonathanknowles commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -526,11 +528,17 @@ instance IsServerError ErrBalanceTx where
+                , "bootstrap witnesses."
+                , "\n\n"
+                , "The candidate tx is:"
+                , T.pack (show candidateTx)
```
Perhaps here we could also create a rich JSON error object instead of trying to embed all the information in a single string!

Anviking replied:
Done as part of #3766

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs:
```diff
@@ -278,16 +296,71 @@ instance Buildable (PartialTx era) where
+
+
+    , canSpendFrom :: Address -> Bool
+    }
```
Hi @Anviking! 

Just curious as to the purpose of the `CoinSelection` type. Is it a way to gather all the dependencies that `balanceTx` needs into a single record?

If so, would be `BalanceTxContext` a good alternative name?

Anviking replied:
>  Is it a way to garther all the dependencies that balanceTx needs into a single record?

No 🤠

Anviking replied:
> Is it a way to gather all the dependencies that balanceTx needs into a single record?

So it's not "*all* the dependencies", but rather the ones relating to how to select inputs from the `UTxO`. Notably this `CoinSelection` type is distinct from things like `PParams` and `ChangeAddressGen`. The latter because we can in theory use a Shelley wallet `UTxO` for inputs, but generate Byron-style change addresses.

I did actually consider combining it with the `UTxO` type into some kind of `data BalancingSource UTxO TxLayer/assumptions about utxo`, but I wasn't sure the name / overhead was good enough.

I'm open to naming suggestions, but I'm sceptical of `BalanceTxContext`, which sounds like something we could just throw all arguments into without meaningful boundaries.

jonathanknowles replied:
@Anviking thanks for you explanation, I appreciate it.

The issue I have with the term `CoinSelection` is that it's already overloaded elsewhere. If I see a type called `CoinSelection`, my first instinct is to think that it represents a selection (singular noun) of coins, or perhaps a _process_ for selecting coins, which is obviously not what this type represents.

I'd propose to choose a name that fits the intent, which is:
> dependencies ... relating to how to select inputs from the UTxO

How about:

- `InputSelectionContext`
- `UTxOSelectionContext`

I think the suffix `Context` is especially good here, as it conveys the idea that objects of this type provide the **_necessary context_** for balanceTx to know how to select inputs from a UTxO set.

Anviking replied:
- `CoinSelector`? 🤔

Anviking replied:
- `CoinSelectionContext`? 🤔 (I mean this is basically the same thing as `{Input,UTxO}SelectionContext`

Anviking replied:
We've gone with `UTxOAssumptions` now as part of the preceding #3811

jonathanknowles reviewed (commented):
Hi @Anviking 
Just some early comments.

Anviking commented on lib/wallet/src/Cardano/Api/Gen.hs:
```diff
@@ -301,10 +280,10 @@ genSlotNo32 = do
+    [ (3, Lovelace . intCast . getNonNegative @Int <$> arbitrary)
+    , (80, choose (1_000_000, 10_000_000_000))
+    , (10, choose (900_000, 2_000_000))
+    , (7, genEncodingBoundaryLovelace)
```
Will revisit all changes to the generators...

Anviking replied:
Dropped for now

Anviking committed:
Refactor balanceTx test setup

Adds property testing of byron wallets.

Unfortunately the tests don't catch the remaining problem of
`estimateKeyWitnessCount` always returning 0 byron witness counts.

The next commit will provide the fix.

Testing will be done with

- Integration tests (when postTransactionOld calls balanceTx)
  - https://input-output.atlassian.net/browse/ADP-2268
- Future `signTx` `estimateKeyWitnessCount` refactoring and testing
  - https://input-output.atlassian.net/browse/ADP-2675
  - https://input-output.atlassian.net/browse/ADP-2676

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2459,6 +2470,136 @@ instance Arbitrary KeyHash where
   arbitrary = StdGenSeed . fromIntegral @Int <$> arbitrary
 
+-- "Wallet"
+data Wallet' = Wallet'
```
Should move this to old location for nicer diff

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2510,9 +2567,10 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
+                let balanceSize = evaluateMinimumFeeDerivedWitSize tx
+                let csSize = coinSelectionEstimatedSize $ intCast n
+                balanceSize === csSize
+                -- >= would suffice, but we can be stronger
```
Renamed to fix name shadowing warning

Anviking committed:
Support byron inputs in estimateKeyWitnessCount

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3928,26 +4025,44 @@ prop_balanceTransactionValid
-        case res of
+        let originalOuts = txOutputs (view #tx partialTx)
+
+        let classifications =
```
Moving these outside the `Right` pattern match. Otherwise the results are heavily biased.

Unisay commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2459,6 +2470,52 @@ instance Arbitrary KeyHash where
+        let
+            firstChangeAddr = fst $ gen s
+        in
+            show $ toLedger firstChangeAddr
```
```suggestion            
            show $ toLedger $ fst $ gen s
```

Anviking committed:
Review suggestions

Co-authored-by: Yura Lazarev <Unisay@users.noreply.github.com>
